[Rokfin] Search & login

P-reducible opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

Add Rokfin site search and login.
Closes #2991

pukkandan commented:
Rebase on master

dirkf commented:
>- [ ] I am the original author of this code and I am willing to release it under Unlicense ??

nyuszika7h commented:
I mean you can't be the author of nothing...

![Screenshot](https://user-images.githubusercontent.com/482367/157425090-4c1b4ee8-da8a-4882-9885-5657f410090f.jpeg)

dirkf commented:
That was going to be the next "??"!

pukkandan commented:
This was split off from #1534. So I assume OP decided to open a PR before cleaning up and committing the actual code to the branch. As long as the code is committed before un-drafting the PR, it's fine

P-reducible commented:
@pukkandan I've checked the "I am the original author ..." box. Does this solve the licensing problem?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +257,64 @@ def _real_extract(self, url):
+            re.finditer(r'<script\s+[^>]*?src\s*=\s*"(?P<path>/static/js/[^">]*)"[^>]*>', notfound_err_page), '')
+        service_url = None
+        services_access_key = None
+        for m in re.finditer(r'(REACT_APP_SEARCH_KEY\s*:\s*"(?P<key>[^"]*)")|(REACT_APP_ENDPOINT_BASE\s*:\s*"(?P<url>[^"]*)")', js):
+            if m.group('url'):
+                service_url = m.group('url') + '/api/as/v1/engines/rokfin-search/search.json'
+            elif m.group('key'):
+                services_access_key = 'Bearer ' + m.group('key')
+            if service_url and services_access_key:
+                return (service_url, services_access_key)
```
1. Why iter through `r're1|re2'` when u can just run 2 `search_regex` on re1 and re2, especially since the code path is split
2. When empty string shouldn't match, use `+`

P-reducible replied:
@pukkandan 

> u can just run 2 search_regex on re1 and re2

Performance. Doing two separate searches is definitely simpler, but it's also slower. These JS files are huge. Why iterate multiple times unless absolutely necessary?

pukkandan replied:
Have you actually profiled it? I doubt there would be any significant difference

pukkandan replied:
actually, it would probably be way faster if you write the regex as: 
```py
r'REACT_APP_(?P<name>SEARCH_KEY|ENDPOINT_BASE)\s*:\s*"(?P<value>[^"]+)'
```
But the real question is whether speed is even a factor here. How long does the regex matching take when compared to the whole extraction process?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +257,64 @@ def _real_extract(self, url):
+
+    def _get_access_credentials(self):
+        notfound_err_page = self._download_webpage('https://rokfin.com/discover', self._SEARCH_KEY, expected_status=404, fatal=False)
+        js = functools.reduce(
+            lambda result, m: result + self._download_webpage(
+                self._BASE_URL + m.group('path'), self._SEARCH_KEY,
+                note='Downloading JavaScript file', fatal=False) or '',
+            re.finditer(r'<script\s+[^>]*?src\s*=\s*"(?P<path>/static/js/[^">]*)"[^>]*>', notfound_err_page), '')
```
it is overkill to download ALL javascript just to match 2 vars. Try and trim it down, or atleast download the js lazily (ie, download one js, search vars, download next...).

Also, it may be worth caching the url and key to disk. (using `self._downloader.cache`)

pukkandan commented:
I have made a force-push to clean up the commit history. You don't need to keep merging unless there is a conflict. It doesn't accomplish anything

P-reducible commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +257,64 @@ def _real_extract(self, url):
+            self._service_url, self._service_access_key = self._get_access_credentials()
+        QUERY_DATA = {'query': query, 'page': {'size': self._ENTRIES_PER_PAGE, 'current': 1}}
+        total_pages = None
+        while True:
+            search_results = self._download_json(
+                self._service_url, self._SEARCH_KEY, headers={'authorization': self._service_access_key},
+                data=json.dumps(QUERY_DATA).encode('utf-8'), encoding='utf-8',
+                note='Downloading search results page %d%s' % (QUERY_DATA['page']['current'], (' of ~%d') % (total_pages,) if total_pages and total_pages >= QUERY_DATA['page']['current'] else ''),
+                fatal=False) or {}
+            QUERY_DATA['page']['current'] += 1
+            total_pages = traverse_obj(search_results, ('meta', 'page', 'total_pages'), expected_type=int_or_none)
+            yield from get_video_data(search_results)
+            if not search_results.get('results'):
+                return
```
The check `total_pages >= page_number` is not essential. But removing it creates instances like
```
$ yt-dlp.sh -s rkfnsearch10000:"\"mollie james\""
...................................................................
[rokfin:search] rkfnsearch: Downloading search results page 2 of ~1
```

pukkandan replied:
imo, that's fine. But I'll leave it to your discretion

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+    _search_engine_access_key = None
+
+    def _real_initialize(self):
+        self._search_engine_access_url = self._downloader.cache.load(self._CACHE_SECTION_NAME, 'url')
+        self._search_engine_access_key = self._downloader.cache.load(self._CACHE_SECTION_NAME, 'key')
+        if not self._search_engine_access_url or not self._search_engine_access_key:
+            self._search_engine_access_url, self._search_engine_access_key = self._get_search_engine_access_credentials()
```
```suggestion
        self._access_url, self._access_key = self._downloader.cache.load(
            self.ie_key(), 'auth') or self._get_access_credentials()
```
PS: We are in the search class. No need to put `search_engine` everywhere!

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+
+    def _search_results(self, query):
+        def get_video_data(metadata):
+            for search_result in metadata.get('results', []):
```
```suggestion
            for search_result in metadata.get('results') or []:
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+    def _search_results(self, query):
+        def get_video_data(metadata):
+            for search_result in metadata.get('results', []):
+                video_id_ind, video_type = self._TYPES.get(traverse_obj(search_result, ('content_type', 'raw')), (None, None))
```
what's `ind`?

P-reducible replied:
`video_id_ind` should have been called `video_id_key`.

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+                video_id_ind, video_type = self._TYPES.get(traverse_obj(search_result, ('content_type', 'raw')), (None, None))
+                video_id = traverse_obj(search_result, video_id_ind, expected_type=int_or_none)
+                if not video_id or not video_type:
+                    self.write_debug(msg=f'skipping {search_result}')
```
```suggestion
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+                    self.write_debug(msg=f'skipping {search_result}')
+                    continue
+                yield self.url_result(url=f'{self._BASE_URL}/{video_type}/{video_id}')
+        if not query:
+            return
```
?

P-reducible replied:
What's your question here?

pukkandan replied:
Why this code?
```py
        if not query:
            return
```

P-reducible replied:
If the query is empty, there is nothing left to do.

pukkandan replied:
this code will never be triggered. besides, extractors shouldn't return None (this doesn't raise an error only due to legacy reasons)

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+                yield self.url_result(url=f'{self._BASE_URL}/{video_type}/{video_id}')
+        if not query:
+            return
+        query_data = {'query': query, 'page': {'size': 100}}
+        total_pages = None
+        for page_number in itertools.count(1):
+            query_data['page']['current'] = page_number
+            search_results = self._run_search_query(
+                data=query_data,
```
```suggestion
        total_pages = None
        for page_number in itertools.count(1):
            search_results = self._run_search_query(
                data={'query': query, 'page': {'size': 100, 'current': page_number}},
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+            query_data['page']['current'] = page_number
+            search_results = self._run_search_query(
+                data=query_data,
+                note='Downloading search results page %d%s' % (page_number, format_field(total_pages, template=' of ~%d') if total_pages and total_pages >= page_number else ''),
+                errnote='Unable to download search results page %d%s' % (page_number, format_field(total_pages, template=' of ~%d') if total_pages and total_pages >= page_number else ''))
```
```suggestion
                note='Downloading search results page %d%s' % (page_number, f' of ~{total_pages}' if total_pages and total_pages >= page_number else ''))
```
OR
```suggestion
                note='Downloading search results page %d%s' % (page_number, format_field(total_pages, template=' of ~%d')))
```
If you are keeping the `total_pages >= page_number`, format_field is useless hee

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+            encoding='utf-8', data=data_bytes, headers={'authorization': self._search_engine_access_key})
+        if search_results is not False:
+            return search_results
+        self.write_debug('updating access credentials')
```
`_get_search_engine_access_credentials` will show necessary messages
```suggestion
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+        if search_results is not False:
+            return search_results
+        self.write_debug('updating access credentials')
+        self._search_engine_access_url = self._search_engine_access_key = None
+        self._downloader.cache.store(self._CACHE_SECTION_NAME, 'url', None)
+        self._downloader.cache.store(self._CACHE_SECTION_NAME, 'key', None)
```
unnecessary

P-reducible replied:
In the latest version, these three lines became
```
self._db_url = self._db_access_key = None
self._downloader.cache.store(self.ie_key(), 'auth', None)
```
(Wiping out the invalid URL/key pair from cache). Does this solve the issue?

pukkandan replied:
But why do u have to to wipe it from cache? You are immediately replacing these values in the next lines

pukkandan replied:
In that case, an error is raised, no?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+        self._downloader.cache.store(self._CACHE_SECTION_NAME, 'url', None)
+        self._downloader.cache.store(self._CACHE_SECTION_NAME, 'key', None)
+        self._search_engine_access_url, self._search_engine_access_key = self._get_search_engine_access_credentials()
+        return self._download_json(
+            self._search_engine_access_url, self._SEARCH_KEY, note=note, errnote=errnote, fatal=False,
+            encoding='utf-8', data=data_bytes, headers={'authorization': self._search_engine_access_key}) or {}
```
DRY

You can instead do:
```py
for attempt in range(2):
    search_results = self._download_json(..., errnote=False, ...)
    if search_results:
        return search_results
    if not attempt:
        ... = self._get_search_engine_access_credentials()
raise ExtractorError(...)
```

P-reducible replied:
Can you clarify what we are accomplishing here?

pukkandan replied:
https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+                return
+
+    def _run_search_query(self, data, note, errnote):
+        data_bytes = json.dumps(data).encode('utf-8')
+        search_results = self._download_json(
+            self._search_engine_access_url, self._SEARCH_KEY, note=note, errnote=errnote, fatal=False,
+            encoding='utf-8', data=data_bytes, headers={'authorization': self._search_engine_access_key})
```
```suggestion
        search_results = self._download_json(
            self._search_engine_access_url, self._SEARCH_KEY, note=note, errnote=False, fatal=False,
            encoding='utf-8', data=json.dumps(data).encode('utf-8'), headers={'authorization': self._access_key})
```
also, is `encoding='utf-8'` needed?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+            encoding='utf-8', data=data_bytes, headers={'authorization': self._search_engine_access_key}) or {}
+
+    def _get_search_engine_access_credentials(self):
+        notfound_err_page = self._download_webpage('https://rokfin.com/discover', self._SEARCH_KEY, expected_status=404, fatal=False)
```
why fatal=False?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+        notfound_err_page = self._download_webpage('https://rokfin.com/discover', self._SEARCH_KEY, expected_status=404, fatal=False)
+        js_content = ''
+        search_engine_access_url = search_engine_access_key = None
+        for js_file_path in re.finditer(r'<script\s+[^>]*?src\s*=\s*"(?P<path>/static/js/[^">]*)"[^>]*>', notfound_err_page):
```
```suggestion
        for js_file_path in re.finditer(r'<script[^>]+src\s*=\s*"(?P<path>/static/js/[^">]+)"', notfound_err_page):
```

P-reducible replied:
`<script[^>]+src\s*=` matches, e.g., `<scripttttttsrc=` which is not our intention. There should at least be one space following `<script`, right?

dirkf replied:
You can use `\s` or `\b`:
```
r'<script\s[^>]*?\bsrc\s*=\s*"(?P<path>/static/js/[^">]+)"[^>]*>'
r'<script\b[^>]*?\bsrc\s*=\s*"(?P<path>/static/js/[^">]+)"[^>]*>'
```
No point using `\s+` as `[^>]` matches whitespace. Also, you presumably want the `path` group not to match just `/static/js/`, and to be the value of a `src` attribute rather than say `notasrc`; the above patterns match, eg, `... data-src="/static/js/abc.js" ...`: if you only want `... src=...` to match:
```
r'<script\b[^>]*?(?<!\w|-)src\s*=\s*"(?P<path>/static/js/[^">]+)"[^>]*>'
```
 Finally, you could also allow for single quotes:
```
r'''<script\b[^>]*?\bsrc\s*=\s*(?P<q>"|')(?P<path>/static/js/(?:(?!(?P=q))[^>])+)(?P=q)[^>]*>'''
```

pukkandan replied:
> <script[^>]+src\s*= matches, e.g., <scripttttttsrc= which is not our intention.

That is not valid html, but sure - you can keep the `\s`. It's more important to fix the `?` and `*`

P-reducible replied:
@pukkandan I don't see anything wrong with the question mark here. Should the corrected version be

`
'<script\s+[^>]*?\s+src\s*=\s*"(?P<path>/static/js/[^">]+)"\s+[^>]*>'
`

? This would match `'<script attr1 = "val1" attr2 src  = "/static/js/file.js" attr3="val3"  >'` but not `'<script attr1 = "val1" attr2src  = "/static/js/file.js" attr3="val3"  >'` and not `'<script attr1 = "val1" attr2 src  = "/static/js/" attr3="val3"  >'`.

If you want, I can remove the trailing `'\s+[^>]*>'` part which I only added for robustness sake.

pukkandan replied:
>  I don't see anything wrong with the question mark here.

And "nothing wrong" is good enough reason to add useless operators?

> '<script\s+[^>]*?\s+src\s*=\s*"(?P<path>/static/js/[^">]+)"\s+[^>]*>'

https://regex101.com/r/s8Vekc/1

```suggestion
        for js_file_path in re.finditer(r'<script\b[^>]*\ssrc\s*=\s*"(?P<path>/static/js/[^">]+)"', notfound_err_page):
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+        js_content = ''
+        search_engine_access_url = search_engine_access_key = None
+        for js_file_path in re.finditer(r'<script\s+[^>]*?src\s*=\s*"(?P<path>/static/js/[^">]*)"[^>]*>', notfound_err_page):
+            js_content += self._download_webpage(
+                self._BASE_URL + js_file_path.group('path'), self._SEARCH_KEY,
+                note='Downloading JavaScript file', fatal=False) or ''
```
```suggestion
            js_content = self._download_webpage(
                self._BASE_URL + js_file_path.group('path'), video_id,
                note='Downloading JavaScript file', fatal=False) or ''
```
set `video_id` to `None` in real_initialize, but must be provided when running from real_extract

P-reducible replied:
The intention behind the `+=` is to concatenate JS files. I assume that a variable assignment (`foo=1`) may **theoretically** be split over multiple files.

pukkandan replied:
Why do you need to concatenate since you already search the previous files?

P-reducible replied:
Because the string being matched may **theoretically** be broken up into multiple pieces, each in its own JS file. The files must be put together for the search to succeed.

pukkandan replied:
how can a dict be split b/w 2 js files? 😕

P-reducible replied:
Why not (in theory)? Does the JS standard prohibit this?

Lesmiscore replied:
In theory JS files can be split, but never at the middle of any object (dict in Python).

P-reducible replied:
Thanks a lot, @Lesmiscore. By any chance, can you send a link?

Lesmiscore replied:
You can't obviously do that like Python can't! https://replit.com/@Lesmiscore/test-js

P-reducible replied:
@pukkandan 

> set video_id to None in real_initialize, but must be provided when running from real_extract

Can you clarify?

pukkandan replied:
When using `_download_webpage` etc from `_real_initialize`, pass `video_id=None`

But from `_real_extract`, it should be an id of the video/playlist (Not necessarily the final `video_id`, just something that can help identify it)

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+            if not search_engine_access_url:
+                search_engine_access_url = self._search_regex(
+                    r'REACT_APP_ENDPOINT_BASE\s*:\s*"(?P<url>[^"]*)"', js_content,
+                    name='Search engine URL', default=None, fatal=False, group='url')
```
```suggestion
                    name='Search engine URL', default=None, group='url')
```
default implies non-fatal. same below

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+                search_engine_access_url = url_or_none(search_engine_access_url + '/api/as/v1/engines/rokfin-search/search.json') if search_engine_access_url else None
+            if not search_engine_access_key:
+                search_engine_access_key = self._search_regex(
+                    r'REACT_APP_SEARCH_KEY\s*:\s*"(?P<key>[^"]*)"', js_content,
```
```suggestion
                    r'REACT_APP_SEARCH_KEY\s*:\s*"(?P<key>[^"]+)"', js_content,
```
why does everyone use * when the match must not be empty 😫

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+                note='Downloading JavaScript file', fatal=False) or ''
+            if not search_engine_access_url:
+                search_engine_access_url = self._search_regex(
+                    r'REACT_APP_ENDPOINT_BASE\s*:\s*"(?P<url>[^"]*)"', js_content,
```
```suggestion
                    r'REACT_APP_ENDPOINT_BASE\s*:\s*"(?P<url>[^"]+)"', js_content,
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+                search_engine_access_key = self._search_regex(
+                    r'REACT_APP_SEARCH_KEY\s*:\s*"(?P<key>[^"]*)"', js_content,
+                    name='Search engine access key', default=None, fatal=False, group='key')
+                search_engine_access_key = ('Bearer ' + search_engine_access_key) if search_engine_access_key else None
```
```suggestion
                search_engine_access_key = f'Bearer {search_engine_access_key}' if search_engine_access_key else None
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+                    name='Search engine access key', default=None, fatal=False, group='key')
+                search_engine_access_key = ('Bearer ' + search_engine_access_key) if search_engine_access_key else None
+            if search_engine_access_url and search_engine_access_key:
+                self._downloader.cache.store(self._CACHE_SECTION_NAME, 'url', search_engine_access_url)
+                self._downloader.cache.store(self._CACHE_SECTION_NAME, 'key', search_engine_access_key)
```
see above comment about cache

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+            if search_engine_access_url and search_engine_access_key:
+                self._downloader.cache.store(self._CACHE_SECTION_NAME, 'url', search_engine_access_url)
+                self._downloader.cache.store(self._CACHE_SECTION_NAME, 'key', search_engine_access_key)
+                return (search_engine_access_url, search_engine_access_key)
```
```suggestion
                return search_engine_access_url, search_engine_access_key
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +422,91 @@ def _real_extract(self, url):
+            js_content += self._download_webpage(
+                self._BASE_URL + js_file_path.group('path'), self._SEARCH_KEY,
+                note='Downloading JavaScript file', fatal=False) or ''
+            if not search_engine_access_url:
+                search_engine_access_url = self._search_regex(
+                    r'REACT_APP_ENDPOINT_BASE\s*:\s*"(?P<url>[^"]*)"', js_content,
+                    name='Search engine URL', default=None, fatal=False, group='url')
+                search_engine_access_url = url_or_none(search_engine_access_url + '/api/as/v1/engines/rokfin-search/search.json') if search_engine_access_url else None
+            if not search_engine_access_key:
+                search_engine_access_key = self._search_regex(
+                    r'REACT_APP_SEARCH_KEY\s*:\s*"(?P<key>[^"]*)"', js_content,
+                    name='Search engine access key', default=None, fatal=False, group='key')
+                search_engine_access_key = ('Bearer ' + search_engine_access_key) if search_engine_access_key else None
```
See my previous comment about grouping the regex together

P-reducible replied:
The initial regex runs slightly faster, though I don't have the numbers in front of me. The margin, however, isn't big enough and makes little difference, given that the process is IO-bound.

pukkandan reviewed (changes requested):
will review login after search is finalized

pukkandan commented:
I noticed you requested a re-review, but many of my earlier comments are still unaddressed?

P-reducible commented:
@pukkandan GitHub hides some of the conversations when it thinks there are too many of them. 11 of them were hidden in my case. Not showing these conversations by default makes them easy to miss, and I assumed them to have already been resolved.

pukkandan commented:
Yes, it also hides some conversations as "outdated" when you change some surrounding code. I recommend uncollapsing and double-checking the comments when you think you are done

pukkandan reviewed (changes requested):
I just realized that the search extractor doesn't have any tests! Please add a test

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +418,82 @@ def _real_extract(self, url):
             f'{channel_id}-{tab}', f'{channel_name} - {tab.title()}', str_or_none(channel_info.get('description')))
+
+
+# E.g.: rkfnsearch5:"\"zelenko\"" or rkfnsearch5:"\"mollie james\""
+class RokfinSearchIE(SearchInfoExtractor):
+    IE_DESC = 'Rokfin Search'
+    _TYPES = {
+        'video': (('id', 'raw'), 'post'),
+        'audio': (('id', 'raw'), 'post'),
+        'stream': (("content_id", 'raw'), 'stream'),
+        'dead_stream': (('content_id', 'raw'), 'stream'),
+        'stack': (('content_id', 'raw'), 'stack'),
+    }
+    IE_NAME = 'rokfin:search'
+    _SEARCH_KEY = 'rkfnsearch'
```
```suggestion
class RokfinSearchIE(SearchInfoExtractor):
    IE_NAME = 'rokfin:search'
    IE_DESC = 'Rokfin Search'
    _SEARCH_KEY = 'rkfnsearch'

    _TYPES = {
        'video': (('id', 'raw'), 'post'),
        'audio': (('id', 'raw'), 'post'),
        'stream': (("content_id", 'raw'), 'stream'),
        'dead_stream': (('content_id', 'raw'), 'stream'),
        'stack': (('content_id', 'raw'), 'stack'),
    }
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +418,82 @@ def _real_extract(self, url):
+    }
+    IE_NAME = 'rokfin:search'
+    _SEARCH_KEY = 'rkfnsearch'
+    _NETRC_MACHINE = False
```
why?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -210,6 +372,8 @@ class RokfinChannelIE(RokfinPlaylistBaseIE):
             'description': 'md5:bb622b1bca100209b91cd685f7847f06',
         },
     }]
+    IE_NAME = 'rokfin:channel'
```
move to top

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -210,6 +372,8 @@ class RokfinChannelIE(RokfinPlaylistBaseIE):
         },
     }]
+    IE_NAME = 'rokfin:channel'
+    _NETRC_MACHINE = False
```
why?

P-reducible replied:
Without `NETRC_MACHINE = False`, line 542 in common.py prints

> WARNING: [rokfin:channel] Login with password is not supported for this website. Use --cookies-from-browser or --cookies for the authentication. See  https://github.com/ytdl-org/youtube-dl#how-do-i-pass-cookies-to-youtube-dl  for how to manually pass cookies

pukkandan replied:
Well, the warning is correct, no? Why do you think it can be suppressed?

P-reducible replied:
Does Rokfin not support logins with passwords? Did I miss something?

pukkandan replied:
`rokfin:channel` does not. If you can use the same login code for it, move it (along with `NETRC_MACHINE` to a base class)

P-reducible replied:
Ok, now I can see what the warning says. But it's still really misleading to say "Login with password is not supported for this **website**".

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -191,6 +351,8 @@ class RokfinStackIE(RokfinPlaylistBaseIE):
             'id': '271',
         },
     }]
+    IE_NAME = 'rokfin:stack'
+    _NETRC_MACHINE = False
```
same

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        LOGIN_PAGE_URL = 'https://secure.rokfin.com/auth/realms/rokfin-web/protocol/openid-connect/auth?client_id=web&redirect_uri=https%3A%2F%2Frokfin.com%2Ffeed&response_mode=fragment&response_type=code&scope=openid'
+        AUTHENTICATION_URL_REGEX_STEP_1 = r'\<form\s+[^>]+action\s*=\s*"(?P<authentication_point_url>https://secure\.rokfin\.com/auth/realms/rokfin-web/login-actions/authenticate\?[^"]+)"[^>]*>'
+
+        # In OpenID terms (https://openid.net/specs/), this program acts as a (public) Client (a.k.a. Client Application),
+        # the User Agent, and the Relying Party (RP) simultaneously, with the Identity Provider (IDP) being secure.rokfin.com.
+        # Rokfin uses KeyCloak (https://www.keycloak.org) as the OpenID implementation of choice.
+        #
+        # ---------------------------- CODE FLOW AUTHORIZATION ----------------------------
+        # https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth (Sec. 3.1.1)
+        #
+        # Authentication phase:
+        #
+        # Step 1: preparation
```
```suggestion
        # https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth (Sec. 3.1)
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        resp_body = self._download_webpage(
+            authentication_point_url, None, note='logging in', fatal=False, expected_status=404, encoding='utf-8',
+            data=urlencode_postdata({'username': username, 'password': password, 'rememberMe': 'off', 'credentialId': ''}))
+        # rememberMe=off resets the session when yt-dlp exits:
+        # https://web.archive.org/web/20220218003425/https://wjw465150.gitbooks.io/keycloak-documentation/content/server_admin/topics/login-settings/remember-me.html
```
yt-dlp is not a browser!

pukkandan replied:
> rememberMe=off resets the session when yt-dlp exits:

This comment is wrong (since yt-dlp is not a browser). You can leave rememberme in the code, but delete the comment

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+
+        return access_token
+
+    def _clear_cookies(self):
+        self._downloader.cookiejar.clear, lambda f: f(domain='secure.rokfin.com')
```
is this a mistake?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        authentication_point_url = unescapeHTML(re.search(AUTHENTICATION_URL_REGEX_STEP_1, login_page).group('authentication_point_url'))
+        if authentication_point_url is None:
+            self.report_warning('login failed unexpectedly: Rokfin extractor must be updated')
+            self._clear_cookies()
```
unless you have a good reason, dont clear cookies

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        if not login_page:
+            return
+        authentication_point_url = unescapeHTML(re.search(AUTHENTICATION_URL_REGEX_STEP_1, login_page).group('authentication_point_url'))
+        if authentication_point_url is None:
+            self.report_warning('login failed unexpectedly: Rokfin extractor must be updated')
```
seems unnecessary. The next `_download_webpage` will raise appropriate error

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+            self._clear_cookies()
+            return
+
+        # Step 2 & 3: authentication
```
```suggestion
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        # rememberMe=off resets the session when yt-dlp exits:
+        # https://web.archive.org/web/20220218003425/https://wjw465150.gitbooks.io/keycloak-documentation/content/server_admin/topics/login-settings/remember-me.html
+        if not self._authentication_active():
+            self._clear_cookies()
```
same as above

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        # https://web.archive.org/web/20220218003425/https://wjw465150.gitbooks.io/keycloak-documentation/content/server_admin/topics/login-settings/remember-me.html
+        if not self._authentication_active():
+            self._clear_cookies()
+            self.report_warning('login failed' + (': invalid username or password.' if type(resp_body) is str and re.search(r'invalid\s+username\s+or\s+password', resp_body, re.IGNORECASE) else ''))
+            return
```
raise error

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+            self.report_warning('login failed' + (': invalid username or password.' if type(resp_body) is str and re.search(r'invalid\s+username\s+or\s+password', resp_body, re.IGNORECASE) else ''))
+            return
+
+        # Authorization phase:
+        #
+        # Steps 4-7:
+        access_mgmt_tokens = self._get_OAuth_tokens()
+        if not access_mgmt_tokens:
+            self._logout()
+            return
+
+        # Validation phase (step 8) skipped, i.e.
+        #
+        # (1) this extractor does not implement client-side ID-token validation;
+        # (2) Rokfin does not supply Subject Identifier.
```
```suggestion
        access_mgmt_tokens = self._get_OAuth_tokens()
        if not access_mgmt_tokens:
            self._logout()
            return
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        # Steps 4-7:
+        access_mgmt_tokens = self._get_OAuth_tokens()
+        if not access_mgmt_tokens:
+            self._logout()
```
is there a reason logout is needed?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
                 return
 
+    def _perform_login(self, username, password):
+        LOGIN_PAGE_URL = 'https://secure.rokfin.com/auth/realms/rokfin-web/protocol/openid-connect/auth?client_id=web&redirect_uri=https%3A%2F%2Frokfin.com%2Ffeed&response_mode=fragment&response_type=code&scope=openid'
+        AUTHENTICATION_URL_REGEX_STEP_1 = r'\<form\s+[^>]+action\s*=\s*"(?P<authentication_point_url>https://secure\.rokfin\.com/auth/realms/rokfin-web/login-actions/authenticate\?[^"]+)"[^>]*>'
```
only used once, can be inlined

pukkandan replied:
```suggestion
        LOGIN_PAGE_URL = 'https://secure.rokfin.com/auth/realms/rokfin-web/protocol/openid-connect/auth?client_id=web&redirect_uri=https%3A%2F%2Frokfin.com%2Ffeed&response_mode=fragment&response_type=code&scope=openid'
        AUTHENTICATION_URL_REGEX_STEP_1 = r'\<form\s+[^>]+action\s*=\s*"(?P<authentication_point_url>https://secure\.rokfin\.com/auth/realms/rokfin-web/login-actions/authenticate\?[^"]+)"'
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+    def _authentication_active(self):
+        current_time_utc = datetime.utcnow().timestamp()
+        SESSION_COOKIE_NAMES = {'KEYCLOAK_IDENTITY', 'KEYCLOAK_IDENTITY_LEGACY', 'KEYCLOAK_SESSION', 'KEYCLOAK_SESSION_LEGACY'}
+        return set([cookie.name for cookie in self._downloader.cookiejar if (cookie.name in SESSION_COOKIE_NAMES)
```
Why are you looking at ALL cookies in the jar?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        SESSION_COOKIE_NAMES = {'KEYCLOAK_IDENTITY', 'KEYCLOAK_IDENTITY_LEGACY', 'KEYCLOAK_SESSION', 'KEYCLOAK_SESSION_LEGACY'}
+        return set([cookie.name for cookie in self._downloader.cookiejar if (cookie.name in SESSION_COOKIE_NAMES)
+                    and (cookie.name not in ('KEYCLOAK_SESSION', 'KEYCLOAK_SESSION_LEGACY')
+                         or cookie.expires > current_time_utc)]) == SESSION_COOKIE_NAMES
```
expired cookies dont stay in the cookiejar. So u just need to check if the cookies exist

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        # https://web.archive.org/web/20220215040021/https://keycloak.discourse.group/t/revoking-or-invalidating-an-authorization-token/1032
+
+    def _authentication_active(self):
+        current_time_utc = datetime.utcnow().timestamp()
+        SESSION_COOKIE_NAMES = {'KEYCLOAK_IDENTITY', 'KEYCLOAK_IDENTITY_LEGACY', 'KEYCLOAK_SESSION', 'KEYCLOAK_SESSION_LEGACY'}
+        return set([cookie.name for cookie in self._downloader.cookiejar if (cookie.name in SESSION_COOKIE_NAMES)
+                    and (cookie.name not in ('KEYCLOAK_SESSION', 'KEYCLOAK_SESSION_LEGACY')
+                         or cookie.expires > current_time_utc)]) == SESSION_COOKIE_NAMES
```
I think this is what u want:
```suggestion
        return bool(traverse_obj(
            self._get_cookies('https://www.rokfin.com'),
            'KEYCLOAK_IDENTITY', 'KEYCLOAK_IDENTITY_LEGACY', 'KEYCLOAK_SESSION', 'KEYCLOAK_SESSION_LEGACY'))
```

P-reducible replied:
The proposed solution didn't work. I have now made this function into a one-liner:

```
def _authentication_active(self):
    return not ({'KEYCLOAK_IDENTITY', 'KEYCLOAK_IDENTITY_LEGACY', 'KEYCLOAK_SESSION', 'KEYCLOAK_SESSION_LEGACY'} - set([cookie.name for cookie in self._downloader.cookiejar]))
```

pukkandan replied:
That is incorrect! Please share account details with me (pukkandan.ytdlp@gmail.com) so I can test why it does not work and fix it

P-reducible replied:
I can create a fresh account for you if you like, but it won't be as much hassle if you do it yourself. Creating a Rokfin account is trivial and free of charge. By the way, it's not `self._get_cookies('https://www.rokfin.com')`. It's `self._get_cookies('https://secure.rokfin.com')`.

pukkandan replied:
okay. Assuming only email is needed to make an account, I'll make one

pukkandan replied:
actually, I think I figured out the issue. You wanted to check if ALL the keys are present, while I only checked if ANY of the keys are present

```py
def _authentication_active(self):
    return all(
        key in self._get_cookies('https://secure.rokfin.com')
        for key in {'KEYCLOAK_IDENTITY', 'KEYCLOAK_IDENTITY_LEGACY', 'KEYCLOAK_SESSION', 'KEYCLOAK_SESSION_LEGACY'})
```
OR
```py
def _authentication_active(self):
    return not (
        {'KEYCLOAK_IDENTITY', 'KEYCLOAK_IDENTITY_LEGACY', 'KEYCLOAK_SESSION', 'KEYCLOAK_SESSION_LEGACY'}
        - set(self._get_cookies('https://secure.rokfin.com')))
```

P-reducible replied:
Unfortunately, neither of them works.

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        authorization_hdr_val = try_get(self._access_mgmt_tokens, lambda tokens: tokens['token_type'] + ' ' + tokens['access_token'])
+        if authorization_hdr_val:
+            headers['authorization'] = authorization_hdr_val
+        json_string, urlh = self._download_webpage_handle(
+            url_or_request, video_id, note='Downloading JSON metadata' + (' [logged in]' if 'authorization' in headers else ''),
+            errnote='Unable to download JSON metadata' + (' [logged in]' if 'authorization' in headers else ''),
+            headers=headers, query=query, expected_status=401)  # 401=Unauthorized
```
```suggestion
        json_string, urlh = self._download_webpage_handle(url_or_request, video_id),
            headers=headers, query=query, expected_status=401)
```
you don't need to add to each line you are logged in. It can be surmised from earlier messages

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+            headers=headers, query=query, expected_status=401)  # 401=Unauthorized
+        if not authorization_hdr_val or urlh.code != 401 or self._access_mgmt_tokens.get('refresh_token') is None:
+            return self._parse_json(json_string, video_id)
+        del headers['authorization']
+        del self._access_mgmt_tokens['access_token']
+        del self._access_mgmt_tokens['token_type']
+        try:
+            del self._access_mgmt_tokens['expires_in']
+        except KeyError:
+            pass
```
why?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+            self._download_json(
+                self._TOKEN_DISTRIBUTION_POINT_URL_STEP_6_7, video_id, note='Restoring lost authorization',
+                errnote='Failed to restore authorization', fatal=False,
+                data=urlencode_postdata({'grant_type': 'refresh_token', 'refresh_token': self._access_mgmt_tokens.get('refresh_token'), 'client_id': 'web'})) or {})
```
line too long

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+                errnote='Failed to restore authorization', fatal=False,
+                data=urlencode_postdata({'grant_type': 'refresh_token', 'refresh_token': self._access_mgmt_tokens.get('refresh_token'), 'client_id': 'web'})) or {})
+        self.write_debug(f'Updated tokens: {self._access_mgmt_tokens.keys()}')
+        if {'access_token', 'expires_in', 'refresh_expires_in', 'refresh_token', 'token_type', 'id_token', 'not-before-policy', 'session_state', 'scope'} - self._access_mgmt_tokens.keys():
+            username, password = self._get_login_info()
+            self._logout()
+            self._perform_login(username, password)
```
Have you observed re-logins helping with this site? If not, this is just overcomplicating things

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+            username, password = self._get_login_info()
+            self._logout()
+            self._perform_login(username, password)
+        authorization_hdr_val = try_get(self._access_mgmt_tokens, lambda tokens: tokens['token_type'] + ' ' + tokens['access_token'])
+        if authorization_hdr_val:
+            headers['authorization'] = authorization_hdr_val
+        return self._download_json(
+            url_or_request, video_id, note='Downloading JSON metadata' + (' [logged in]' if 'authorization' in headers else ''),
+            errnote='Unable to download JSON metadata' + (' [logged in]' if 'authorization' in headers else ''),
+            headers=headers, query=query)
```
DRY

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+    def _get_OAuth_tokens(self):
+        PARTIAL_USER_CONSENT_URL_STEP_4_5 = urllib.parse.urlparse('https://secure.rokfin.com/auth/realms/rokfin-web/protocol/openid-connect/auth?client_id=web&redirect_uri=https%3A%2F%2Frokfin.com%2Fsilent-check-sso.html&response_mode=fragment&response_type=code&scope=openid&prompt=none')
+
+        # ---------------------------- CODE FLOW AUTHORIZATION ----------------------------
+        # https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth (Sec. 3.1.1)
+
+        # Steps 4 & 5 authorize yt-dlp to act on user's behalf.
+
+        # random_str() came from https://www.rokfin.com/static/js/2.*.chunk.js
```
```suggestion
        # From https://www.rokfin.com/static/js/2.*.chunk.js
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+
+        user_consent_url_step_4_5 = PARTIAL_USER_CONSENT_URL_STEP_4_5._replace(query=urllib.parse.urlencode((lambda d: d.update(state=random_str(), nonce=random_str()) or d)(dict(urllib.parse.parse_qsl(PARTIAL_USER_CONSENT_URL_STEP_4_5.query))))).geturl()
+
+        # By making this HTTP request, the user authorizes yt-dlp to act on behalf of the user:
```
```suggestion
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        user_consent_url_step_4_5 = PARTIAL_USER_CONSENT_URL_STEP_4_5._replace(query=urllib.parse.urlencode((lambda d: d.update(state=random_str(), nonce=random_str()) or d)(dict(urllib.parse.parse_qsl(PARTIAL_USER_CONSENT_URL_STEP_4_5.query))))).geturl()
+
+        # By making this HTTP request, the user authorizes yt-dlp to act on behalf of the user:
+        urlh = (self._download_webpage_handle(
+            user_consent_url_step_4_5, None, note='granting user authorization', errnote='user authorization rejected by Rokfin', fatal=False, encoding='utf-8') or (None, None))[1]
```
1. why is this non-fatal?
2. If you don't need the webpage content, use _request_webpage
3. _request/download_webpage accepts a `query` param. No need of the messy URL manipulations above

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+            return codes[0] if codes else None
+
+        authorization_code = parse_authorization_code(urlh)
+        if not authorization_code:
+            self.write_debug('authorization failed: missing authorization code')
+            return None
```
shouldn't this be a warning/error?

P-reducible replied:
Yes, it should. But are `--abort-on-error`/`--no-abort-on-error` relevant here? I know that, technically, these two only apply to download errors. Even so, can they be used in the current situation?

pukkandan replied:
No. Generally, we raise error for user+pass login. But if you think it makes sense to attempt extraction even with failed login, you can just give a warning instead

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+            self._TOKEN_DISTRIBUTION_POINT_URL_STEP_6_7, None, note='getting access credentials', fatal=False, encoding='utf-8',
+            data=urlencode_postdata({'code': authorization_code, 'grant_type': 'authorization_code', 'client_id': 'web', 'redirect_uri': 'https://rokfin.com/silent-check-sso.html'})) or {}
+
+        # Usage: --extractor-args "rokfin:print-user-info"  # For testing by those who don't have premium access.
+        if self._configuration_arg('print_user_info'):
+            self.to_screen(self._download_json('https://prod-api-v2.production.rokfin.com/api/v2/user/me', None, note='obtaining user info', errnote='could not obtain user info', fatal=False, headers={'authorization': access_token['token_type'] + ' ' + access_token['access_token']}))
```
dont leave debugging code in...

P-reducible replied:
In your opinion, this snippet is not important for future testing without paid subscription (as the comment says)?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +169,157 @@ def _get_comments(self, video_id):
+        # Steps 6 & 7 request and acquire ID Token & Access Token:
+        access_token = self._download_json(
+            self._TOKEN_DISTRIBUTION_POINT_URL_STEP_6_7, None, note='getting access credentials', fatal=False, encoding='utf-8',
+            data=urlencode_postdata({'code': authorization_code, 'grant_type': 'authorization_code', 'client_id': 'web', 'redirect_uri': 'https://rokfin.com/silent-check-sso.html'})) or {}
```
too long

P-reducible commented:
@pukkandan Hopefully, I have addressed all the issues except those related to the `_authentication_active()` function which you seem to be working on. Perhaps you can merge RokfinSearchIE?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +324,85 @@ def _real_extract(self, url):
+        'dead_stream': (('content_id', 'raw'), 'stream'),
+        'stack': (('content_id', 'raw'), 'stack'),
+    }
+    _TESTS = [{
+        'url': 'rkfnsearch5:\"zelenko\"',  # The quotes are intentional.
+        'playlist_count': 5,
+        'info_dict': {
+            'id': '\"zelenko\"',
+            'title': '\"zelenko\"',
+        }
+    }]
```
if this is meant to be the equivalent of typing `"zelenko"` in the site, it should be
```suggestion
    _TESTS = [{
        'url': 'rkfnsearch5:"zelenko"',
        'playlist_count': 5,
        'info_dict': {
            'id': '"zelenko"',
            'title': '"zelenko"',
        }
    }]
```

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +168,66 @@ def _get_comments(self, video_id):
+
+        # https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth (Sec. 3.1)
+        login_page = self._download_webpage(LOGIN_PAGE_URL, None, note='loading login page', errnote='error loading login page')
+        authentication_point_url = unescapeHTML(re.search(AUTHENTICATION_URL_REGEX_STEP_1, login_page).group('authentication_point_url'))
```
_search_regex

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +168,66 @@ def _get_comments(self, video_id):
+        resp_body = self._download_webpage(
+            authentication_point_url, None, note='logging in', fatal=False, expected_status=404,
+            data=urlencode_postdata({'username': username, 'password': password, 'rememberMe': 'off', 'credentialId': ''}))
+        if not self._authentication_active():
+            raise ExtractorError('login failed' + (': invalid username and/or password.' if type(resp_body) is str and re.search(r'invalid\s+username\s+or\s+password', resp_body, re.IGNORECASE) else ''), expected=True)
```
1. line too long
2. Is expected=True even when not wrong password?

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -160,6 +168,66 @@ def _get_comments(self, video_id):
+            urlh = self._request_webpage(
+                'https://secure.rokfin.com/auth/realms/rokfin-web/protocol/openid-connect/auth', None,
+                note='granting user authorization', errnote='user authorization rejected by Rokfin',
+                query={'client_id': 'web', 'redirect_uri': 'https://rokfin.com/silent-check-sso.html',
+                       'response_mode': 'fragment', 'response_type': 'code', 'scope': 'openid',
+                       'prompt': 'none'})
```
read "Trailing parentheses" section in code style

same issue elsewhere

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -21,6 +26,9 @@
 class RokfinIE(InfoExtractor):
     _VALID_URL = r'https?://(?:www\.)?rokfin\.com/(?P<id>(?P<type>post|stream)/\d+)'
+    _NETRC_MACHINE = 'rokfin'
+    _TOKEN_DISTRIBUTION_POINT_URL_STEP_6_7 = 'https://secure.rokfin.com/auth/realms/rokfin-web/protocol/openid-connect/token'
```
best to move inside the login function imo

P-reducible replied:
I don't know how I could move `_TOKEN_DISTRIBUTION_POINT_URL_STEP_6_7` into `_perform_login` since `_TOKEN_DISTRIBUTION_POINT_URL_STEP_6_7` is being shared by  `_perform_login` and `_download_json_using_access_token`.

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -254,3 +324,85 @@ def _real_extract(self, url):
+                db_url = self._search_regex(
+                    r'REACT_APP_ENDPOINT_BASE\s*:\s*"(?P<url>[^"]+)"', js_content,
+                    name='Search engine URL', default=None, group='url')
+                db_url = url_or_none(db_url + '/api/as/v1/engines/rokfin-search/search.json') if db_url else None
```
```suggestion
                db_url = url_or_none(f'{db_url}/api/as/v1/engines/rokfin-search/search.json')
```

pukkandan reviewed (changes requested):
See previous unresolved comments

> except those related to the _authentication_active() function which you seem to be working on

I'll re-check this when everything else is done

P-reducible commented:
> See previous unresolved comments

Which ones? Didn't I address them all?

pukkandan commented:
I've marked all addressed ones as "resolved". Anything that isn't marked as such is pending

P-reducible commented:
This looks like a bug in the merged section:

```
$ yt-dlp.sh https://rokfin.com/stream/8892
[Rokfin] stream/8892: Downloading webpage
ERROR: [Rokfin] stream/8892: Unable to download webpage: HTTP Error 404: Not Found (caused by <HTTPError 404: 'Not Found'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```

The user should not have to file a bug report when the video does not exist.

pukkandan commented:
pls verify my changes

pukkandan commented on yt_dlp/extractor/rokfin.py:
```diff
@@ -159,6 +166,78 @@ def _get_comments(self, video_id):
+                'redirect_uri': 'https://rokfin.com/silent-check-sso.html'
+            }))
+
+    def _authentication_active(self):  # TODO
+        return not ({'KEYCLOAK_IDENTITY', 'KEYCLOAK_IDENTITY_LEGACY', 'KEYCLOAK_SESSION', 'KEYCLOAK_SESSION_LEGACY'} - set([cookie.name for cookie in self._downloader.cookiejar]))
```
note to self: TODO
